File: src/AutoContext.Mcp.Web/src/features/workspace-server/workspace-server-client.ts
import { connect } from 'node:net';

import type { McpToolsRequest } from './mcp-tools/mcp-tools-request.js';
import type { McpToolsResponse } from './mcp-tools/mcp-tools-response.js';
import type { LogRequest } from './logging/log-request.js';

export class WorkspaceServerClient {
    private readonly pipePath: string | undefined;
    private readonly source: string;

    constructor(pipeName?: string, source?: string) {
        this.source = source ?? 'Unknown';

        if (pipeName) {
            this.pipePath = process.platform === 'win32'
                ? `\\\\.\\pipe\\${pipeName}`
                : `/tmp/CoreFxPipe_${pipeName}`;
        }
    }

    /**
     * Resolves tool modes and EditorConfig data for a batch of MCP tools
     * via the `mcp-tools` workspace service endpoint.
     */
    async resolveTools(request: McpToolsRequest): Promise<McpToolsResponse | undefined> {
        if (!this.pipePath) {
            return undefined;
        }

        const response = await this.sendPipeRequest<McpToolsResponse>({
            type: 'mcp-tools',
            tools: request.tools,
            'file-path': request.filePath,
            'editorconfig-keys': request['editorconfig-keys'],
        });

        return response.tools ? response : undefined;
    }

    /**
     * Sends a log message to the workspace service for centralized output.
     * Falls back to local stderr when the pipe is unavailable.
     */
    async sendLog(level: string, message: string): Promise<void> {
        if (!this.pipePath) {
            // No pipe configured — write directly to stderr (expected when running standalone).
            process.stderr.write(`[${this.source}] ${message}\n`);
            return;
        }

        try {
            const request: LogRequest = {
                type: 'log',
                source: this.source,
                level,
                message,
            };
            await this.sendPipeRequest(request);
        } catch (ex) {
            // Pipe failed — log the reason and the original message to local stderr.
            const reason = ex instanceof Error ? ex.message : String(ex);
            process.stderr.write(`[${this.source}] Failed to send log via workspace server: ${reason}\n`);
            process.stderr.write(`[${this.source}] ${message}\n`);
        }
    }

    /**
     * Low-level pipe protocol: sends a length-prefixed JSON request and
     * returns the parsed response object.
     */
    private sendPipeRequest<T>(request: object): Promise<T> {
        if (!this.pipePath) {
            return Promise.resolve({} as T);
        }

        const json = JSON.stringify(request);
        const requestBytes = Buffer.from(json, 'utf8');
        const header = Buffer.alloc(4);
        header.writeInt32LE(requestBytes.length, 0);

        const pipePath = this.pipePath;

        return new Promise((resolve) => {
            const chunks: Buffer[] = [];
            let received = 0;

            const socket = connect(pipePath, () => {
                socket.write(header);
                socket.write(requestBytes);
            });

            socket.on('data', (chunk: Buffer) => {
                chunks.push(chunk);
                received += chunk.length;

                // Need at least 4 bytes for the length header.
                if (received < 4) {
                    return;
                }

                // Read length from the first chunk if it's large enough,
                // otherwise concat once to get the header.
                const headerBuf = chunks[0]!.length >= 4
                    ? chunks[0]!
                    : Buffer.concat(chunks);

                const payloadLength = headerBuf.readInt32LE(0);

                // Need the full payload.
                if (received < 4 + payloadLength) {
                    return;
                }

                const buf = chunks.length === 1 ? chunks[0]! : Buffer.concat(chunks);
                const payloadBytes = buf.subarray(4, 4 + payloadLength);
                socket.destroy();

                try {
                    const response = JSON.parse(payloadBytes.toString('utf8')) as T;
                    resolve(response);
                } catch {
                    resolve({} as T);
                }
            });

            socket.on('error', () => resolve({} as T));
            socket.on('timeout', () => { socket.destroy(); resolve({} as T); });
            socket.setTimeout(5000);
        });
    }
}

File: src/AutoContext.Mcp.Shared/WorkspaceServer/WorkspaceServerClient.cs
namespace AutoContext.Mcp.Shared.WorkspaceServer;

using System.Buffers.Binary;
using System.IO.Pipes;
using System.Text.Json;

using AutoContext.Mcp.Shared.WorkspaceServer.Logging;
using AutoContext.Mcp.Shared.WorkspaceServer.McpTools;

/// <summary>
/// Named pipe client for the <c>AutoContext.WorkspaceServer</c> service process.
/// </summary>
/// <remarks>
/// Initializes a new instance of the <see cref="WorkspaceServerClient"/> class.
/// </remarks>
/// <param name="pipeName">Named pipe used to connect to the workspace service, or <see langword="null"/> when not available.</param>
/// <param name="source">Server identity included in log messages (e.g. "DotNet", "Git").</param>
public sealed class WorkspaceServerClient(string? pipeName = null, string? source = null)
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.KebabCaseLower,
    };

    private readonly string? _pipeName = pipeName;
    private readonly string _source = source ?? "Unknown";

    /// <summary>
    /// Resolves tool modes and EditorConfig data for a batch of MCP tools
    /// via the <c>mcp-tools</c> workspace service endpoint.
    /// </summary>
    internal async Task<McpToolsResponse?> ResolveToolsAsync(McpToolsRequest request, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(_pipeName))
        {
            return null;
        }

        var requestBytes = JsonSerializer.SerializeToUtf8Bytes(request, JsonOptions);

        using var client = new NamedPipeClientStream(".", _pipeName, PipeDirection.InOut, PipeOptions.Asynchronous);
        await client.ConnectAsync(5000, ct).ConfigureAwait(false);

        await WriteMessageAsync(client, requestBytes, ct).ConfigureAwait(false);

        var responseBytes = await ReadMessageAsync(client, ct).ConfigureAwait(false);

        if (responseBytes is null || responseBytes.Length == 0)
        {
            return null;
        }

        return JsonSerializer.Deserialize<McpToolsResponse>(responseBytes, JsonOptions);
    }

    /// <summary>
    /// Sends a log message to the workspace service for centralized output.
    /// Falls back to local stderr when the pipe is unavailable.
    /// </summary>
    internal async Task SendLogAsync(string level, string message, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(_pipeName))
        {
            // No pipe configured — write directly to stderr (expected when running standalone).
            await Console.Error.WriteLineAsync($"[{_source}] {message}").ConfigureAwait(false);
            return;
        }

        try
        {
            var request = new LogRequest(_source, level, message);
            var requestBytes = JsonSerializer.SerializeToUtf8Bytes(request, JsonOptions);

            using var client = new NamedPipeClientStream(".", _pipeName, PipeDirection.InOut, PipeOptions.Asynchronous);
            await client.ConnectAsync(5000, ct).ConfigureAwait(false);

            await WriteMessageAsync(client, requestBytes, ct).ConfigureAwait(false);

            // Protocol requires reading the response to complete the cycle.
            await ReadMessageAsync(client, ct).ConfigureAwait(false);
        }
        catch (IOException ex)
        {
            // Pipe failed — log the reason and the original message to local stderr.
            await Console.Error.WriteLineAsync(
                $"[{_source}] Failed to send log via workspace server: {ex.Message}").ConfigureAwait(false);
            await Console.Error.WriteLineAsync($"[{_source}] {message}").ConfigureAwait(false);
        }
        catch (TimeoutException ex)
        {
            await Console.Error.WriteLineAsync(
                $"[{_source}] Failed to send log via workspace server: {ex.Message}").ConfigureAwait(false);
            await Console.Error.WriteLineAsync($"[{_source}] {message}").ConfigureAwait(false);
        }
    }

    private static async Task WriteMessageAsync(Stream stream, byte[] payload, CancellationToken ct = default)
    {
        var message = new byte[4 + payload.Length];
        BinaryPrimitives.WriteInt32LittleEndian(message, payload.Length);
        payload.CopyTo(message.AsSpan(4));

        await stream.WriteAsync(message, ct).ConfigureAwait(false);
        await stream.FlushAsync(ct).ConfigureAwait(false);
    }

    private static async Task<byte[]?> ReadMessageAsync(Stream stream, CancellationToken ct = default)
    {
        var header = new byte[4];

        if (!await ReadExactAsync(stream, header, ct).ConfigureAwait(false))
        {
            return null;
        }

        var length = BinaryPrimitives.ReadInt32LittleEndian(header);

        if (length <= 0)
        {
            return [];
        }

        var payload = new byte[length];

        return await ReadExactAsync(stream, payload, ct).ConfigureAwait(false) ? payload : null;
    }

    private static async Task<bool> ReadExactAsync(Stream stream, byte[] buffer, CancellationToken ct = default)
    {
        var offset = 0;

        while (offset < buffer.Length)
        {
            var read = await stream.ReadAsync(
                buffer.AsMemory(offset, buffer.Length - offset), ct).ConfigureAwait(false);

            if (read == 0)
            {
                return false;
            }

            offset += read;
        }

        return true;
    }
}
